server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -480,16 +485,16 @@ private void internalRecoverFromStore(IndexShard indexShard, ActionListener<Void
                 indexShard.openEngineAndRecoverFromTranslog();
                 indexShard.getEngine().fillSeqNoGaps(indexShard.getPendingPrimaryTerm());
                 indexShard.finalizeRecovery();
-                indexShard.postRecovery("post recovery from shard_store", ActionListener.runBefore(listener, store::decRef));
-                triggeredPostRecovery = true;
-            } catch (EngineException | IOException e) {
-                listener.onFailure(new IndexShardRecoveryException(shardId, "failed to recover from gateway", e));
-            } finally {
-                if (triggeredPostRecovery == false) {
-                    store.decRef();
+                indexShard.postRecovery("post recovery from shard_store", l);
+            })
+
+            .addListener(ActionListener.runBefore(outerListener.delegateResponse((l, e) -> {
+                if (e instanceof IndexShardRecoveryException) {
+                    l.onFailure(e);
+                } else {
+                    l.onFailure(new IndexShardRecoveryException(shardId, "failed to recover from gateway", e));
```

Comment:
This is a slight change (IMO improvement) in the failure behaviour: _all_ exceptions now become and `IndexShardRecoveryException` and therefore mention the shard ID.